storage/remote/write_handler.go

```diff
@@ -34,11 +39,68 @@ import (
 )
 
 const (
-	RemoteWriteVersionHeader       = "X-Prometheus-Remote-Write-Version"
-	RemoteWriteVersion1HeaderValue = "0.1.0"
-	RemoteWriteVersion2HeaderValue = "2.0"
+	RemoteWriteVersionHeader        = "X-Prometheus-Remote-Write-Version"
+	RemoteWriteVersion1HeaderValue  = "0.1.0"
+	RemoteWriteVersion20HeaderValue = "2.0"
 )
 
+func rwHeaderNameValues(rwFormat config.RemoteWriteFormat) map[string]string {
+	// Return the correct remote write header name/values based on provided rwFormat
+	ret := make(map[string]string, 1)
+
+	switch rwFormat {
+	case Version1:
+		ret[RemoteWriteVersionHeader] = RemoteWriteVersion1HeaderValue
+	case Version2:
+		// We need to add the supported protocol definitions in order:
+		tuples := make([]string, 0, 2)
+		// Add 2.0;snappy
+		tuples = append(tuples, RemoteWriteVersion20HeaderValue+";snappy")
+		// Add default 0.1.0
+		tuples = append(tuples, RemoteWriteVersion1HeaderValue)
+		ret[RemoteWriteVersionHeader] = strings.Join(tuples, ",")
+	}
+	return ret
+}
+
+type writeHeadHandler struct {
+	logger log.Logger
+
+	remoteWrite20HeadRequests prometheus.Counter
+
+	// Experimental feature, new remote write proto format
+	// The handler will accept the new format, but it can still accept the old one
+	rwFormat config.RemoteWriteFormat
+}
+
+func NewWriteHeadHandler(logger log.Logger, reg prometheus.Registerer, rwFormat config.RemoteWriteFormat) http.Handler {
+	h := &writeHeadHandler{
+		logger:   logger,
+		rwFormat: rwFormat,
+		remoteWrite20HeadRequests: prometheus.NewCounter(prometheus.CounterOpts{
+			Namespace: "prometheus",
+			Subsystem: "api",
+			Name:      "remote_write_20_head_requests",
+			Help:      "The number of remote write 2.0 head requests.",
+		}),
+	}
+	if reg != nil {
+		reg.MustRegister(h.remoteWrite20HeadRequests)
+	}
+	return h
+}
+
+func (h *writeHeadHandler) ServeHTTP(w http.ResponseWriter, r *http.Request) {
+	// Send a response to the HEAD request based on the format supported
```

Comment:
Should this be comment for ServeHTTP instead (on top not below)

Reply:
Yes, fixing in next PR: https://github.com/prometheus/prometheus/pull/13921